File: src/year2023/day6/day6.py
#!/usr/bin/env python3
# [Day 6: Wait For It](https://adventofcode.com/2023/day/6)

import atexit
import math
import sys
import time
from pathlib import Path

filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()
if "--elapsed" in sys.argv:
    sys.argv.remove("--elapsed")
    start_time_ns = time.time_ns()
    atexit.register(lambda: print(f"elapsed: {(time.time_ns() - start_time_ns) / 1_000_000}ms"))


def wins(times, distances):
    result = 1

    for t, d in zip(times, distances):
        # win = 0
        # for hold in range(1, t):
        #     if hold * (t - hold) > d:
        #         win += 1

        # hold * (t - hold) > d ⇔ a < hold < b
        a = (t - (t * t - 4 * d) ** 0.5) / 2
        b = (t + (t * t - 4 * d) ** 0.5) / 2

        # 1 ≤ hold < t
        a = max(1, math.floor(a))
        b = min(t, math.ceil(b))

        win = b - a - 1

        result *= win

    return result


# parse input
lines = data.splitlines()

times = list(map(int, lines[0].removeprefix("Time:").split()))
distances = list(map(int, lines[1].removeprefix("Distance:").split()))

# part 1
print(wins(times, distances))

# part 2
times = (int(lines[0].removeprefix("Time:").replace(" ", "")),)
distances = (int(lines[1].removeprefix("Distance:").replace(" ", "")),)
print(wins(times, distances))

File: src/year2023/day6/day6.rs
//! [Day 6: Wait For It](https://adventofcode.com/2023/day/6)

struct Puzzle {
    time: String,
    distance: String,
}

impl Puzzle {
    fn new(data: &str) -> Self {
        let mut lines = data.lines();

        Self {
            time: lines
                .next()
                .unwrap()
                .strip_prefix("Time:")
                .unwrap()
                .to_string(),
            distance: lines
                .next()
                .unwrap()
                .strip_prefix("Distance:")
                .unwrap()
                .to_string(),
        }
    }

    // fn win(t: u64, d: u64) -> u64 {
    //     // nota: see below for the elegant math solution 🤓
    //     let mut win = 0;
    //     for hold in 1..t {
    //         let speed = hold;
    //         let remaining = t - hold;
    //         let travelled = speed * remaining;
    //         if travelled > d {
    //             win += 1;
    //         }
    //     }
    //     win
    // }

    fn win(t: u64, d: u64) -> u64 {
        // solve hold * (t - hold) > d
        // e.g find (a,b) such that a < hold < b

        let t = u128::from(t); // use u128 to avoid overflow
        let d = u128::from(d);

        let sqrt_disc = (t * t - 4 * d).isqrt(); // Δ = t² - 4 d
        let mut a = (t - sqrt_disc) / 2; //  ⌊ (t - √Δ) / 2 ⌋
        let mut b = (t + sqrt_disc).div_ceil(2); // ⌈ (t + √Δ) / 2 ⌉

        if a * (t - a) > d {
            a -= 1;
        }
        if b * (t - b) > d {
            b += 1;
        }

        u64::try_from(b - a - 1).unwrap()
    }

    /// Solve part one.
    fn part1(&self) -> u64 {
        let time = self
            .time
            .split_whitespace()
            .map(|x| x.parse::<u64>().unwrap());
        let distance = self
            .distance
            .split_whitespace()
            .map(|x| x.parse::<u64>().unwrap());

        let mut result = 1;
        for (t, d) in time.zip(distance) {
            result *= Self::win(t, d);
        }
        result
    }

    /// Solve part two.
    fn part2(&self) -> u64 {
        let time = self.time.replace(' ', "").parse::<u64>().unwrap();
        let distance = self.distance.replace(' ', "").parse::<u64>().unwrap();

        Self::win(time, distance)
    }
}

/// # Panics
#[must_use]
pub fn solve(data: &str) -> (u64, u64) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn test01() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part1(), 288);
    }

    #[test]
    fn test02() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part2(), 71503);
    }
}
